source/distributed/tria.cc

```diff
@@ -1124,10 +1124,545 @@ namespace
+                unpack_callback(dealii_cell,
+                                cell_status,
+                                boost::make_iterator_range(cit_dest_fix,
+                                                           cit_dest_fix +
+                                                             size));
+                break;
+
+              case parallel::distributed::Triangulation<dim,
+                                                        spacedim>::CELL_REFINE:
+                unpack_callback(dealii_cell->parent(),
+                                cell_status,
+                                boost::make_iterator_range(cit_dest_fix,
+                                                           cit_dest_fix +
+                                                             size));
+                break;
+
+              case parallel::distributed::Triangulation<dim,
+                                                        spacedim>::CELL_INVALID:
+                // Skip this cell.
+                break;
+
+              default:
+                Assert(false, ExcInternalError());
+                break;
+            }
+        }
+    }
+
+
+
+    template <int dim, int spacedim>
+    void
+    Triangulation<dim, spacedim>::DataTransfer::save(
+      const typename dealii::internal::p4est::types<dim>::forest
+        *         parallel_forest,
+      const char *filename) const
+    {
+      Assert(cumulative_sizes_fix.size() > 0,
+             ExcMessage("No data has been packed!"));
+
+      const std::string fname_fixed = std::string(filename) + "_fixed.data";
+
+      // ----- copied -----
+      // from DataOutInterface::write_vtu_parallel
+      // TODO: write general MPIIO interface
+      int myrank, nproc;
+      int ierr = MPI_Comm_rank(mpi_communicator, &myrank);
+      AssertThrowMPI(ierr);
+      ierr = MPI_Comm_size(mpi_communicator, &nproc);
+      AssertThrowMPI(ierr);
+
+      MPI_Info info;
+      ierr = MPI_Info_create(&info);
+      AssertThrowMPI(ierr);
+      MPI_File fh;
+      ierr = MPI_File_open(mpi_communicator,
+                           const_cast<char *>(fname_fixed.c_str()),
+                           MPI_MODE_CREATE | MPI_MODE_WRONLY,
+                           info,
+                           &fh);
+      AssertThrowMPI(ierr);
+
+      ierr = MPI_File_set_size(fh, 0); // delete the file contents
+      AssertThrowMPI(ierr);
+      // this barrier is necessary, because otherwise others might already
+      // write while one core is still setting the size to zero.
+      ierr = MPI_Barrier(mpi_communicator);
+      AssertThrowMPI(ierr);
+      ierr = MPI_Info_free(&info);
+      AssertThrowMPI(ierr);
+      // ------------------
+
+      // Check if number of processors is lined up with p4est partitioning.
+      Assert(myrank < parallel_forest->mpisize, ExcInternalError());
+
+      // Write cumulative sizes to file.
+      if (myrank == 0)
```

Comment:
make it clear in the comment that only processor zero does this, since the `cumulative_sizes_fix(ed)` are the same on all processors